<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Background - Basic Tutorial for Genome Wide Association Analysis</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Background";
    var mkdocs_page_input_path = "prsice.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Genome Wide Association Analysis</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../QC/">1. Quality Control of GWAS Data</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../popstrat/">2. Population Stratification</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../cal_prs/">3. Association Analyses</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing association results</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="" href="../assocdive.md">5. Deep dive into underpinnings of associations</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../liftover/">liftover</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Genome Wide Association Analysis</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Background</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/prsice.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="background">Background<a class="headerlink" href="#background" title="Permanent link">&para;</a></h1>
<p>PRSice-2 is one of the dedicated PRS programs which automates many of the steps from the previous page that used a sequence of PLINK functions (plus some QC steps). 
On this page you will run a PRS analysis using PRSice-2, which implements the standard C+T method.</p>
<h2 id="obtaining-prsice-2">Obtaining PRSice-2<a class="headerlink" href="#obtaining-prsice-2" title="Permanent link">&para;</a></h2>
<p><code>PRSice-2</code> can be downloaded from:</p>
<table>
<thead>
<tr>
<th>Operating System</th>
<th align="center">Link</th>
</tr>
</thead>
<tbody>
<tr>
<td>Linux 64-bit</td>
<td align="center"><a href="https://github.com/choishingwan/PRSice/releases/download/2.3.3/PRSice_linux.zip">v2.3.3</a></td>
</tr>
<tr>
<td>OS X 64-bit</td>
<td align="center"><a href="https://github.com/choishingwan/PRSice/releases/download/2.3.3/PRSice_mac.zip">v2.3.3</a></td>
</tr>
</tbody>
</table>
<p>and can be directly used after extracting the file. </p>
<p>In this tutorial, you will only need <code>PRSice.R</code> and <code>PRSice_XXX</code> where XXX is the operation system</p>
<h2 id="required-data">Required Data<a class="headerlink" href="#required-data" title="Permanent link">&para;</a></h2>
<p>This analysis assumes that you have the following files (or you can download it from <a href="https://drive.google.com/file/d/1x_G0Gxk9jFMY-PMqwtg6-vdEyUPp5p5u/view?usp=sharing">here</a>): </p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post QC base data file. While PRSice-2 can automatically apply most filtering on the base file, it cannot remove duplicated SNPs</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">This file contains the genotype data that passed the QC steps</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the list of SNPs that passed the QC steps</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the QC steps</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype data of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.cov</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.eigenvec</strong></td>
<td align="center">This file contains the principal components (PCs) of the samples</td>
</tr>
</tbody>
</table>
<h2 id="running-prs-analysis">Running PRS analysis<a class="headerlink" href="#running-prs-analysis" title="Permanent link">&para;</a></h2>
<p>To run PRSice-2 we need a single covariate file, and therefore our covariate file and PCs file should be combined. This can be done with <code>R</code> as follows:</p>
<div class="tabbed-set" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">without data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">covariate &lt;- read.table("EUR.cov", header=T)
pcs &lt;- read.table("EUR.eigenvec", header=F)
colnames(pcs) &lt;- c("FID","IID", paste0("PC",1:6))
cov &lt;- merge(covariate, pcs, by=c("FID", "IID"))
write.table(cov,"EUR.covariate", quote=F, row.names=F)
q()</code></pre>
</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">with data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">library(data.table)
covariate &lt;- fread("EUR.cov")
pcs &lt;- fread("EUR.eigenvec", header=F)
colnames(pcs) &lt;- c("FID","IID", paste0("PC",1:6))
cov &lt;- merge(covariate, pcs)
fwrite(cov,"EUR.covariate", sep="\t")
q()</code></pre>
</div>
</div>
<p>which generates <strong>EUR.cov</strong>.</p>
<p>PRSice-2 can then be run to obtain the PRS results as follows:</p>
<div class="tabbed-set" data-tabs="2:3"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">Linux</label><div class="tabbed-content">
<pre class="highlight"><code class="language-bash">Rscript PRSice.R \
    --prsice PRSice_linux \
    --base Height.QC.gz \
    --target EUR.QC \
    --binary-target F \
    --pheno EUR.height \
    --cov EUR.covariate \
    --base-maf MAF:0.01 \
    --base-info INFO:0.8 \
    --stat OR \
    --or \
    --out EUR</code></pre>
</div>
<input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><label for="__tabbed_2_2">OS X</label><div class="tabbed-content">
<pre class="highlight"><code class="language-bash">Rscript PRSice.R \
    --prsice PRSice_mac \
    --base Height.QC.gz \
    --target EUR.QC \
    --binary-target F \
    --pheno EUR.height \
    --cov EUR.covariate \
    --base-maf MAF:0.01 \
    --base-info INFO:0.8 \
    --stat OR \
    --or \
    --out EUR</code></pre>
</div>
<input id="__tabbed_2_3" name="__tabbed_2" type="radio" /><label for="__tabbed_2_3">Windows</label><div class="tabbed-content">
<pre class="highlight"><code class="language-bash">Rscript PRSice.R ^
    --prsice PRSice_win64.exe ^
    --base Height.QC.gz ^
    --target EUR.QC ^
    --binary-target F ^
    --pheno EUR.height ^
    --cov EUR.covariate ^
    --base-maf MAF:0.05 ^
    --base-info INFO:0.8 ^
    --stat OR ^
    --or ^
    --out EUR</code></pre>
</div>
</div>
<p>The meaning of the parameters are as follow:</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">prsice</td>
<td align="center">PRSice_xxx</td>
<td align="left">Informs <code>PRSice.R</code> that the location of the PRSice binary</td>
</tr>
<tr>
<td align="center">base</td>
<td align="center">Height.QC.gz</td>
<td align="left">Informs <code>PRSice</code> that the name of the GWAS summary statistic</td>
</tr>
<tr>
<td align="center">target</td>
<td align="center">EUR.QC</td>
<td align="left">Informs <code>PRSice</code> that the input genotype files should have a prefix of <code>EUR.QC</code></td>
</tr>
<tr>
<td align="center">binary-target</td>
<td align="center">F</td>
<td align="left">Indicate if the phenotype of interest is a binary trait. F for no</td>
</tr>
<tr>
<td align="center">pheno</td>
<td align="center">EUR.height</td>
<td align="left">Provide <code>PRSice</code> with the phenotype file</td>
</tr>
<tr>
<td align="center">cov</td>
<td align="center">EUR.covariate</td>
<td align="left">Provide <code>PRSice</code> with the covariate file</td>
</tr>
<tr>
<td align="center">base-maf</td>
<td align="center">MAF:0.05</td>
<td align="left">Filter out SNPs with MAF &lt; 0.05 in the GWAS summary statistics, using information in the <code>MAF</code> column</td>
</tr>
<tr>
<td align="center">base-info</td>
<td align="center">INFO:0.8</td>
<td align="left">Filter out SNPs with INFO &lt; 0.8 in the GWAS summary statistics, using information in the <code>INFO</code> column</td>
</tr>
<tr>
<td align="center">stat</td>
<td align="center">OR</td>
<td align="left">Column name of the column containing the effect size</td>
</tr>
<tr>
<td align="center">or</td>
<td align="center">-</td>
<td align="left">Inform <code>PRSice</code> that the effect size is an Odd Ratio</td>
</tr>
<tr>
<td align="center">out</td>
<td align="center">EUR</td>
<td align="left">Informs <code>PRSice</code> that all output should have a prefix of <code>EUR</code></td>
</tr>
</tbody>
</table>
<p>This will automatically perform "high-resolution scoring" and generate the "best-fit" PRS (in <strong>EUR.best</strong>), with associated plots of the results. 
Users should read Section 4.6 of our paper to learn more about issues relating to overfitting in PRS analyses.  </p>
<details class="note"><summary>Which P-value threshold generates the "best-fit" PRS?</summary><p>0.13995</p>
</details>
<details class="note"><summary>How much phenotypic variation does the "best-fit" PRS explain?</summary><p>0.166117</p>
</details>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
