<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>liftover - Basic Tutorial for Genome Wide Association Analysis</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "liftover";
    var mkdocs_page_input_path = "liftover.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Genome Wide Association Analysis</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../QC/">1. Quality Control of GWAS Data</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../popstrat/">2. Population Stratification</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../assoc/">3. Association Analyses</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing association results</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="" href="../assocdive.md">5. Deep dive into underpinnings of associations</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">liftover</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#installing-ldpred-2">Installing LDpred-2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#required-data">Required Data</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#0-prepare-workspace">0. Prepare workspace</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#2-obtain-hapmap3-snps">2. Obtain HapMap3 SNPs</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#3-load-and-transform-the-summary-statistic-file">3. Load and transform the summary statistic file</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#3-calculate-the-ld-matrix">3. Calculate the LD matrix</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#4-perform-ld-score-regression">4. Perform LD score regression</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#5-calculate-the-null-r2">5. Calculate the null R2</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#7-obtain-model-prs">7. Obtain model PRS</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#using-genome-wide-bed-file">Using Genome wide bed file</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#using-chromosome-separated-bed-files">Using chromosome separated bed files</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#8-get-the-final-performance-of-the-ldpred-models">8. Get the final performance of the LDpred models</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Genome Wide Association Analysis</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>liftover</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/liftover.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="background">Background<a class="headerlink" href="#background" title="Permanent link">&para;</a></h1>
<p><a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred-2</a> is one of the dedicated PRS programs which is an <code>R</code> package that uses a Bayesian approach to polygenic risk scoring.</p>
<h2 id="installing-ldpred-2">Installing LDpred-2<a class="headerlink" href="#installing-ldpred-2" title="Permanent link">&para;</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The script used here is based on LDpred 2 implemented under bigsnpr version 1.4.7</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For more details, please refer to <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred 2's homepage</a></p>
</div>
<p>You can install <code>LDpred</code> and its dependencies in <code>R</code> with the following command:</p>
<pre class="highlight"><code class="language-R">install.packages("remotes")
library(remotes)
remotes::install_github("https://github.com/privefl/bigsnpr.git")</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For mac users, you might need to follow the guide <a href="https://thecoatlessprofessor.com/programming/cpp/openmp-in-r-on-os-x/">here</a> to be able to install LDpred2</p>
</div>
<h2 id="required-data">Required Data<a class="headerlink" href="#required-data" title="Permanent link">&para;</a></h2>
<p>We assume that you have the following files (or you can download it from <a href="https://drive.google.com/file/d/1x_G0Gxk9jFMY-PMqwtg6-vdEyUPp5p5u/view?usp=sharing">here</a>):</p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post-QCed summary statistic</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">The genotype file after performing some basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the SNPs that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.cov</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.eigenvec</strong></td>
<td align="center">This file contains the PCs of the samples</td>
</tr>
</tbody>
</table>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>While we do provide a rough guide on how to perform LDpred on bed files separated into individual chromosomes, this script is untested and extra caution is required</p>
</div>
<h3 id="0-prepare-workspace">0. Prepare workspace<a class="headerlink" href="#0-prepare-workspace" title="Permanent link">&para;</a></h3>
<p>On some server, you might need to first use the following code in order to run LDpred with multi-thread</p>
<div class="tabbed-set" data-tabs="1:1"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">prepare workspace and load bigsnpr</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">library(bigsnpr)
options(bigstatsr.check.parallel.blas = FALSE)
options(default.nproc.blas = NULL)</code></pre>
</div>
</div>
<h3 id="1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files<a class="headerlink" href="#1-read-in-the-phenotype-and-covariate-files" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">read in phenotype and covariates</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">library(data.table)
library(magrittr)
phenotype &lt;- fread("EUR.height")
covariate &lt;- fread("EUR.cov")
pcs &lt;- fread("EUR.eigenvec")
# rename columns
colnames(pcs) &lt;- c("FID","IID", paste0("PC",1:6))
# generate required table
pheno &lt;- merge(phenotype, covariate) %&gt;%
    merge(., pcs)</code></pre>
</div>
</div>
<h3 id="2-obtain-hapmap3-snps">2. Obtain HapMap3 SNPs<a class="headerlink" href="#2-obtain-hapmap3-snps" title="Permanent link">&para;</a></h3>
<p>LDpred2 authors recommend restricting the analysis to only the HapMap3 SNPs</p>
<div class="tabbed-set" data-tabs="3:1"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">load HapMap3 SNPs</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">info &lt;- readRDS(url("https://github.com/privefl/bigsnpr/raw/master/data-raw/hm3_variants.rds"))</code></pre>
</div>
</div>
<h3 id="3-load-and-transform-the-summary-statistic-file">3. Load and transform the summary statistic file<a class="headerlink" href="#3-load-and-transform-the-summary-statistic-file" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="4:1"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">Load summary statistic file</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Read in the summary statistic file
sumstats &lt;- bigreadr::fread2("Height.QC.gz") 
# LDpred 2 require the header to follow the exact naming
names(sumstats) &lt;-
    c("chr",
    "pos",
    "rsid",
    "a1",
    "a0",
    "n_eff",
    "beta_se",
    "p",
    "OR",
    "INFO",
    "MAF")
# Transform the OR into log(OR)
sumstats$beta &lt;- log(sumstats$OR)
# Filter out hapmap SNPs
sumstats &lt;- sumstats[sumstats$rsid%in% info$rsid,]</code></pre>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Here, we know the exact ordering of the summary statistics file. 
However, in many cases, the ordering of the summary statistics differ, 
thus one must rename the columns according to their actual ordering</p>
</div>
</div>
</div>
<h3 id="3-calculate-the-ld-matrix">3. Calculate the LD matrix<a class="headerlink" href="#3-calculate-the-ld-matrix" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="5:2"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><label for="__tabbed_5_1">Genome Wide bed file</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Get maximum amount of cores
NCORES &lt;- nb_cores()
# Open a temporary file
tmp &lt;- tempfile(tmpdir = "tmp-data")
on.exit(file.remove(paste0(tmp, ".sbk")), add = TRUE)
# Initialize variables for storing the LD score and LD matrix
corr &lt;- NULL
ld &lt;- NULL
# We want to know the ordering of samples in the bed file 
fam.order &lt;- NULL
# preprocess the bed file (only need to do once for each data set)
snp_readBed("EUR.QC.bed")
# now attach the genotype object
obj.bigSNP &lt;- snp_attach("EUR.QC.rds")
# extract the SNP information from the genotype
map &lt;- obj.bigSNP$map[-3]
names(map) &lt;- c("chr", "rsid", "pos", "a1", "a0")
# perform SNP matching
info_snp &lt;- snp_match(sumstats, map)
# Assign the genotype to a variable for easier downstream analysis
genotype &lt;- obj.bigSNP$genotypes
# Rename the data structures
CHR &lt;- map$chr
POS &lt;- map$pos
# get the CM information from 1000 Genome
# will download the 1000G file to the current directory (".")
POS2 &lt;- snp_asGeneticPos(CHR, POS, dir = ".")
# calculate LD
for (chr in 1:22) {
    # Extract SNPs that are included in the chromosome
    ind.chr &lt;- which(info_snp$chr == chr)
    ind.chr2 &lt;- info_snp$`_NUM_ID_`[ind.chr]
    # Calculate the LD
    corr0 &lt;- snp_cor(
            genotype,
            ind.col = ind.chr2,
            ncores = NCORES,
            infos.pos = POS2[ind.chr2],
            size = 3 / 1000
        )
    if (chr == 1) {
        ld &lt;- Matrix::colSums(corr0^2)
        corr &lt;- as_SFBM(corr0, tmp)
    } else {
        ld &lt;- c(ld, Matrix::colSums(corr0^2))
        corr$add_columns(corr0, nrow(corr))
    }
}
# We assume the fam order is the same across different chromosomes
fam.order &lt;- as.data.table(obj.bigSNP$fam)
# Rename fam order
setnames(fam.order,
        c("family.ID", "sample.ID"),
        c("FID", "IID"))</code></pre>
</div>
<input id="__tabbed_5_2" name="__tabbed_5" type="radio" /><label for="__tabbed_5_2">Chromosome separated bed files</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Get maximum amount of cores
NCORES &lt;- nb_cores()
# Open a temporary file
tmp &lt;- tempfile(tmpdir = "tmp-data")
on.exit(file.remove(paste0(tmp, ".sbk")), add = TRUE)
# Initialize variables for storing the LD score and LD matrix
corr &lt;- NULL
ld &lt;- NULL
# We want to know the ordering of samples in the bed file 
info_snp &lt;- NULL
fam.order &lt;- NULL
for (chr in 1:22) {
    # preprocess the bed file (only need to do once for each data set)
    # Assuming the file naming is EUR_chr#.bed
    snp_readBed(paste0("EUR_chr",chr,".bed"))
    # now attach the genotype object
    obj.bigSNP &lt;- snp_attach(paste0("EUR_chr",chr,".rds"))
    # extract the SNP information from the genotype
    map &lt;- obj.bigSNP$map[-3]
    names(map) &lt;- c("chr", "rsid", "pos", "a1", "a0")
    # perform SNP matching
    tmp_snp &lt;- snp_match(sumstats[sumstats$chr==chr,], map)
    info_snp &lt;- rbind(info_snp, tmp_snp)
    # Assign the genotype to a variable for easier downstream analysis
    genotype &lt;- obj.bigSNP$genotypes
    # Rename the data structures
    CHR &lt;- map$chr
    POS &lt;- map$pos
    # get the CM information from 1000 Genome
    # will download the 1000G file to the current directory (".")
    POS2 &lt;- snp_asGeneticPos(CHR, POS, dir = ".")
    # calculate LD
    # Extract SNPs that are included in the chromosome
    ind.chr &lt;- which(tmp_snp$chr == chr)
    ind.chr2 &lt;- tmp_snp$`_NUM_ID_`[ind.chr]
    # Calculate the LD
    corr0 &lt;- snp_cor(
            genotype,
            ind.col = ind.chr2,
            ncores = NCORES,
            infos.pos = POS2[ind.chr2],
            size = 3 / 1000
        )
    if (chr == 1) {
        ld &lt;- Matrix::colSums(corr0^2)
        corr &lt;- as_SFBM(corr0, tmp)
    } else {
        ld &lt;- c(ld, Matrix::colSums(corr0^2))
        corr$add_columns(corr0, nrow(corr))
    }
    # We assume the fam order is the same across different chromosomes
    if(is.null(fam.order)){
        fam.order &lt;- as.data.table(obj.bigSNP$fam)
    }
}
# Rename fam order
setnames(fam.order,
        c("family.ID", "sample.ID"),
        c("FID", "IID"))</code></pre>
</div>
</div>
<h3 id="4-perform-ld-score-regression">4. Perform LD score regression<a class="headerlink" href="#4-perform-ld-score-regression" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="6:1"><input checked="checked" id="__tabbed_6_1" name="__tabbed_6" type="radio" /><label for="__tabbed_6_1">Perform LD score regression</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">df_beta &lt;- info_snp[,c("beta", "beta_se", "n_eff", "_NUM_ID_")]
ldsc &lt;- snp_ldsc(   ld, 
                    length(ld), 
                    chi2 = (df_beta$beta / df_beta$beta_se)^2,
                    sample_size = df_beta$n_eff, 
                    blocks = NULL)
h2_est &lt;- ldsc[["h2"]]</code></pre>
</div>
</div>
<h3 id="5-calculate-the-null-r2">5. Calculate the null R2<a class="headerlink" href="#5-calculate-the-null-r2" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="7:2"><input checked="checked" id="__tabbed_7_1" name="__tabbed_7" type="radio" /><label for="__tabbed_7_1">Calculate the null R2 (quantitative trait)</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Reformat the phenotype file such that y is of the same order as the 
# sample ordering in the genotype file
y &lt;- pheno[fam.order, on = c("FID", "IID")]
# Calculate the null R2
# use glm for binary trait 
# (will also need the fmsb package to calculate the pseudo R2)
null.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary
null.r2 &lt;- null.model$r.squared</code></pre>
</div>
<input id="__tabbed_7_2" name="__tabbed_7" type="radio" /><label for="__tabbed_7_2">Calculate the null R2 (binary trait)</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">library(fmsb)
# Reformat the phenotype file such that y is of the same order as the 
# sample ordering in the genotype file
y &lt;- pheno[fam.order, on = c("FID", "IID")]
# Calculate the null R2
# use glm for binary trait 
# (will also need the fmsb package to calculate the pseudo R2)
null.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~Sex+", .) %&gt;%
    as.formula %&gt;%
    glm(., data = y, family=binomial) %&gt;%
    summary
null.r2 &lt;- fmsb::NagelkerkeR2(null.model)</code></pre>
</div>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Scripts for binary trait analysis only serve as a reference as we have not simulate any binary traits. 
In addition, Nagelkerke <span class="arithmatex">\(R^2\)</span> is biased when there are ascertainment of samples. For more information, please refer to <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/gepi.21614">this paper</a></p>
</div>
<div class="tabbed-set" data-tabs="8:3"><input checked="checked" id="__tabbed_8_1" name="__tabbed_8" type="radio" /><label for="__tabbed_8_1">infinitesimal model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">beta_inf &lt;- snp_ldpred2_inf(corr, df_beta, h2 = h2_est)</code></pre>
</div>
<input id="__tabbed_8_2" name="__tabbed_8" type="radio" /><label for="__tabbed_8_2">grid model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Prepare data for grid model
p_seq &lt;- signif(seq_log(1e-4, 1, length.out = 17), 2)
h2_seq &lt;- round(h2_est * c(0.7, 1, 1.4), 4)
grid.param &lt;-
    expand.grid(p = p_seq,
            h2 = h2_seq,
            sparse = c(FALSE, TRUE))
# Get adjusted beta from grid model
beta_grid &lt;-
    snp_ldpred2_grid(corr, df_beta, grid.param, ncores = NCORES)</code></pre>
</div>
<input id="__tabbed_8_3" name="__tabbed_8" type="radio" /><label for="__tabbed_8_3">auto model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Get adjusted beta from the auto model
multi_auto &lt;- snp_ldpred2_auto(
    corr,
    df_beta,
    h2_init = h2_est,
    vec_p_init = seq_log(1e-4, 0.9, length.out = NCORES),
    ncores = NCORES
)
beta_auto &lt;- sapply(multi_auto, function(auto)
    auto$beta_est)</code></pre>
</div>
</div>
<h3 id="7-obtain-model-prs">7. Obtain model PRS<a class="headerlink" href="#7-obtain-model-prs" title="Permanent link">&para;</a></h3>
<h4 id="using-genome-wide-bed-file">Using Genome wide bed file<a class="headerlink" href="#using-genome-wide-bed-file" title="Permanent link">&para;</a></h4>
<div class="tabbed-set" data-tabs="9:3"><input checked="checked" id="__tabbed_9_1" name="__tabbed_9" type="radio" /><label for="__tabbed_9_1">infinitesimal model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">if(is.null(obj.bigSNP)){
    obj.bigSNP &lt;- snp_attach("EUR.QC.rds")
}
genotype &lt;- obj.bigSNP$genotypes
# calculate PRS for all samples
ind.test &lt;- 1:nrow(genotype)
pred_inf &lt;- big_prodVec(    genotype,
                            beta_inf,
                            ind.row = ind.test,
                            ind.col = info_snp$`_NUM_ID_`)</code></pre>
</div>
<input id="__tabbed_9_2" name="__tabbed_9" type="radio" /><label for="__tabbed_9_2">grid model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">if(is.null(obj.bigSNP)){
    obj.bigSNP &lt;- snp_attach("EUR.QC.rds")
}
genotype &lt;- obj.bigSNP$genotypes
# calculate PRS for all samples
ind.test &lt;- 1:nrow(genotype)
pred_grid &lt;- big_prodMat(   genotype, 
                            beta_grid, 
                            ind.col = info_snp$`_NUM_ID_`)</code></pre>
</div>
<input id="__tabbed_9_3" name="__tabbed_9" type="radio" /><label for="__tabbed_9_3">auto model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">if(is.null(obj.bigSNP)){
    obj.bigSNP &lt;- snp_attach("EUR.QC.rds")
}
genotype &lt;- obj.bigSNP$genotypes
# calculate PRS for all samples
ind.test &lt;- 1:nrow(genotype)
pred_auto &lt;-
    big_prodMat(genotype,
                beta_auto,
                ind.row = ind.test,
                ind.col = info_snp$`_NUM_ID_`)
# scale the PRS generated from AUTO
pred_scaled &lt;- apply(pred_auto, 2, sd)
final_beta_auto &lt;-
    rowMeans(beta_auto[,
                abs(pred_scaled -
                    median(pred_scaled)) &lt;
                    3 * mad(pred_scaled)])
pred_auto &lt;-
    big_prodVec(genotype,
        final_beta_auto,
        ind.row = ind.test,
        ind.col = info_snp$`_NUM_ID_`)</code></pre>
</div>
</div>
<h4 id="using-chromosome-separated-bed-files">Using chromosome separated bed files<a class="headerlink" href="#using-chromosome-separated-bed-files" title="Permanent link">&para;</a></h4>
<div class="tabbed-set" data-tabs="10:3"><input checked="checked" id="__tabbed_10_1" name="__tabbed_10" type="radio" /><label for="__tabbed_10_1">infinitesimal model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">pred_inf &lt;- NULL
for(chr in 1:22){
    obj.bigSNP &lt;- snp_attach(paste0("EUR_chr",chr,".rds"))
    genotype &lt;- obj.bigSNP$genotypes
    # calculate PRS for all samples
    ind.test &lt;- 1:nrow(genotype)
    # Extract SNPs in this chromosome
    chr.idx &lt;- which(info_snp$chr == chr)
    ind.chr &lt;- info_snp$`_NUM_ID_`[chr.idx]
    tmp &lt;- big_prodVec(genotype,
                            beta_inf[chr.idx],
                            ind.row = ind.test,
                            ind.col = ind.chr)
    if(is.null(pred_inf)){
        pred_inf &lt;- tmp
    }else{
        pred_inf &lt;- pred_inf + tmp
    }
}</code></pre>
</div>
<input id="__tabbed_10_2" name="__tabbed_10" type="radio" /><label for="__tabbed_10_2">grid model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">pred_grid &lt;- NULL
for(chr in 1:22){
    obj.bigSNP &lt;- snp_attach(paste0("EUR_chr",chr,"_.rds"))
    genotype &lt;- obj.bigSNP$genotypes
    # calculate PRS for all samples
    ind.test &lt;- 1:nrow(genotype)
    # Extract SNPs in this chromosome
    chr.idx &lt;- which(info_snp$chr == chr)
    ind.chr &lt;- info_snp$`_NUM_ID_`[chr.idx]

    tmp &lt;- big_prodMat( genotype, 
                        beta_grid[chr.idx], 
                        ind.col = ind.chr)

    if(is.null(pred_grid)){
        pred_grid &lt;- tmp
    }else{
        pred_grid &lt;- pred_grid + tmp
    }
}</code></pre>
</div>
<input id="__tabbed_10_3" name="__tabbed_10" type="radio" /><label for="__tabbed_10_3">auto model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">pred_auto &lt;- NULL
for(chr in 1:22){
    obj.bigSNP &lt;- snp_attach(paste0("EUR_chr",chr,"_.rds"))
    genotype &lt;- obj.bigSNP$genotypes
    # calculate PRS for all samples
    ind.test &lt;- 1:nrow(genotype)
    # Extract SNPs in this chromosome
    chr.idx &lt;- which(info_snp$chr == chr)
    ind.chr &lt;- info_snp$`_NUM_ID_`[chr.idx]

    tmp &lt;-
        big_prodMat(genotype,
                    beta_auto[chr.idx],
                    ind.row = ind.test,
                    ind.col = ind.chr)
    # scale the PRS generated from AUTO
    pred_scaled &lt;- apply(tmp, 2, sd)
    final_beta_auto &lt;-
        rowMeans(beta_auto[chr.idx,
                    abs(pred_scaled -
                        median(pred_scaled)) &lt;
                        3 * mad(pred_scaled)])
    tmp &lt;-
        big_prodVec(genotype,
            final_beta_auto,
            ind.row = ind.test,
            ind.col = ind.chr)
    if(is.null(pred_auto)){
        pred_auto &lt;- tmp
    }else{
        pred_auto &lt;- pred_auto + tmp
    }
}</code></pre>
</div>
</div>
<h3 id="8-get-the-final-performance-of-the-ldpred-models">8. Get the final performance of the LDpred models<a class="headerlink" href="#8-get-the-final-performance-of-the-ldpred-models" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="11:3"><input checked="checked" id="__tabbed_11_1" name="__tabbed_11" type="radio" /><label for="__tabbed_11_1">infinitesimal model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">reg.formula &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~PRS+Sex+", .) %&gt;%
    as.formula
reg.dat &lt;- y
reg.dat$PRS &lt;- pred_inf
inf.model &lt;- lm(reg.formula, dat=reg.dat) %&gt;%
    summary
(result &lt;- data.table(
    infinitesimal = inf.model$r.squared - null.r2,
    null = null.r2
))</code></pre>
</div>
<input id="__tabbed_11_2" name="__tabbed_11" type="radio" /><label for="__tabbed_11_2">grid model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">reg.formula &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~PRS+Sex+", .) %&gt;%
    as.formula
reg.dat &lt;- y
max.r2 &lt;- 0
for(i in 1:ncol(pred_grid)){
    reg.dat$PRS &lt;- pred_grid[,i]
    grid.model &lt;- lm(reg.formula, dat=reg.dat) %&gt;%
        summary  
    if(max.r2 &lt; grid.model$r.squared){
        max.r2 &lt;- grid.model$r.squared
    }
}
(result &lt;- data.table(
    grid = max.r2 - null.r2,
    null = null.r2
))</code></pre>
</div>
<input id="__tabbed_11_3" name="__tabbed_11" type="radio" /><label for="__tabbed_11_3">auto model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">reg.formula &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~PRS+Sex+", .) %&gt;%
    as.formula
reg.dat &lt;- y
reg.dat$PRS &lt;- pred_auto
auto.model &lt;- lm(reg.formula, dat=reg.dat) %&gt;%
    summary
(result &lt;- data.table(
    auto = auto.model$r.squared - null.r2,
    null = null.r2
))</code></pre>
</div>
</div>
<details class="note"><summary>How much phenotypic variation does the PRS from each model explain?</summary><p>Infinitesimal = 0.0100</p>
<p>Grid Model = 0.00180</p>
<p>Auto Model = 0.171</p>
</details>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../plink_visual/" class="btn btn-neutral" title="4. Visualizing association results"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../plink_visual/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
