<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>2. Population Stratification - Basic Tutorial for Genome Wide Association Analysis</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "2. Population Stratification";
    var mkdocs_page_input_path = "popstrat.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Genome Wide Association Analysis</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../QC/">1. Quality Control of GWAS Data</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">2. Population Stratification</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#obtaining-the-1000-genome-reference-data-set">Obtaining the 1000 genome reference data set</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#define-variables-to-be-used-in-this-section">Define Variables to be used in this section</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#qc-on-1000-genomes-data">QC on 1000 Genomes data.</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#extract-the-variants-present-in-1000-genomes-dataset-from-your-dataset">Extract the variants present in 1000 Genomes dataset from your dataset.</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#change-build-on-1000-genomes-data-build-to-match-build-of-hapmap-data">Change build on 1000 Genomes data build to match build of HapMap data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#merge-the-map-and-1000-genomes-data-sets">Merge the Map and 1000 Genomes data sets</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#merge-outdata-with-1000-genomes-data">Merge outdata with 1000 Genomes Data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#perform-mds-and-pca-on-map-ceu-data-anchored-by-1000-genomes-data-using-a-set-of-pruned-snps">Perform MDS and PCA on Map-CEU data anchored by 1000 Genomes data using a set of pruned SNPs</a>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../assoc/">3. Association Analyses</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing association results</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="" href="../assocdive.md">5. Deep dive into underpinnings of associations</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../liftover/">liftover</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../prs/">PRS</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Genome Wide Association Analysis</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>2. Population Stratification</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/popstrat.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="population-stratifacation">Population Stratifacation<a class="headerlink" href="#population-stratifacation" title="Permanent link">&para;</a></h1>
<h2 id="obtaining-the-1000-genome-reference-data-set">Obtaining the 1000 genome reference data set<a class="headerlink" href="#obtaining-the-1000-genome-reference-data-set" title="Permanent link">&para;</a></h2>
<p>We will be using data from the 1000 Genomes Project for the population stratification step.</p>
<p>You can download using the following command </p>
<p><pre class="highlight"><code class="language-bash">wget ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20100804/ALL.2of4intersection.20100804.genotypes.vcf.gz</code></pre>
The data will need to be converted into plink format with unique indentifiers created for each the SNPs with a missing rs-identifier:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You will need <code>plink</code> in this section, which can be download from <a href="https://www.cog-genomics.org/plink/1.9/">here</a>.</p>
<p>Install the program <code>plink</code> and include its location in your PATH directory, which allows us to use <code>plink</code> instead of <code>./plink</code> in the commands below. If PLINK is not in your PATH directory and is instead in your working directory, replace all instances of <code>plink</code> in the tutorial with <code>./plink</code>.</p>
</div>
<pre class="highlight"><code class="language-bash">plink --vcf ALL.2of4intersection.20100804.genotypes.vcf.gz --make-bed --out 1000genomes.genotypes
plink --bfile 1000genomes.genotypes --set-missing-var-ids @:#[b37]\$1,\$2 --make-bed --out 1000genomes_nomissing.genotypes</code></pre>
<h2 id="define-variables-to-be-used-in-this-section">Define Variables to be used in this section<a class="headerlink" href="#define-variables-to-be-used-in-this-section" title="Permanent link">&para;</a></h2>
<p><strong>input files and variables</strong>
<pre class="highlight"><code class="language-bash">FILE_1K=1000genomes_nomissing.genotypes #1000 genomes file
FILE_QC=qcout #file from qc tab
FILE_PRUNEIN=plink.prune.in #snps in approximate linkage equilibrium 
GENO=0.02 #snp missingness filter
INDV=0.02 #individual missingness filter
MAF=0.05 #minor allele frequency filter
HWE_CONTROL=1e-6 #hardy weinburg equilibrium filter</code></pre></p>
<p><strong>define general file tags</strong></p>
<pre class="highlight"><code class="language-bash">sep="_"
tagbed=".bed"
tagbim=".bim"
tagfam=".fam"
tagmap=".map"
taggenome=".genome"
tag_dot_mds=".mds"</code></pre>
<p><strong>tags for filtering 1000genome data</strong>
<pre class="highlight"><code class="language-bash">TAG_1kG="1KG"
TAG_GENO="geno"
OUT1="$TAG_1kG$sep$TAG_GENO$sep$GENO"
TAG_MIND="mind"
OUT2="$OUT1$sep$TAG_MIND$sep$INDV"
TAG_MAF="maf"
OUT3=$OUT2$sep$TAG_MAF$sep$MAF
TAG_extract="extract"
OUT4=$OUT3$sep$TAG_extract
TAG_BUILD="samebuild"
OUT5=$OUT4$sep$TAG_BUILD
TAG_REM="removeproblem"
OUT6=$OUT5$sep$TAG_REM</code></pre></p>
<p><strong>tags for filtering QCed data</strong>
<pre class="highlight"><code class="language-bash">TAG_QC="QCin"
OUT1_QC=$TAG_QC$sep$TAG_extract
OUT2_QC=$OUT1_QC$sep$TAG_REM
tag_euro="euro"
FILEQCEURO=$FILE_QC$sep$tag_euro
TAG_HWE_CONTROL="hwe_control"
OUTQCEURO=$FILEQCEURO$sep$TAG_HWE_CONTROL$sep$HWE_CONTROL
found="founder"
lowcover="unrelated"
OUTCOVER=$lowcover
tag_mds="MDS"
POPSTRATOUT=$OUTCOVER$sep$tag_mds
tag_pca="PCA"
INPCA=$OUTCOVER</code></pre></p>
<h2 id="qc-on-1000-genomes-data">QC on 1000 Genomes data.<a class="headerlink" href="#qc-on-1000-genomes-data" title="Permanent link">&para;</a></h2>
<p><strong>Remove variants based on missing genotype data.</strong>
<pre class="highlight"><code class="language-bash">plink --bfile $FILE_1K --geno $GENO --make-bed --out $OUT1</code></pre>
<strong>Remove individuals based on missing genotype data</strong>
<pre class="highlight"><code class="language-bash">plink --bfile $OUT1 --mind $INDV --allow-no-sex --make-bed --out $OUT2</code></pre>
<strong>Remove variants based on MAF</strong>
<pre class="highlight"><code class="language-bash">plink --bfile $OUT2 --maf $MAF --make-bed --out $OUT3</code></pre>
<strong>Extract the variants present in our dataset from the 1000 genomes dataset</strong>
<pre class="highlight"><code class="language-bash">awk '{print$2}' "$FILE_QC$tagbim"&gt; QCFILE_SNPs.txt
awk '{print$2}' "$OUT3$tagbim"&gt; 1kG_temp.bim
plink --bfile $OUT3 --extract QCFILE_SNPs.txt --make-bed --recode --out $OUT4</code></pre></p>
<h2 id="extract-the-variants-present-in-1000-genomes-dataset-from-your-dataset">Extract the variants present in 1000 Genomes dataset from your dataset.<a class="headerlink" href="#extract-the-variants-present-in-1000-genomes-dataset-from-your-dataset" title="Permanent link">&para;</a></h2>
<p><pre class="highlight"><code class="language-bash">awk '{print$2}' $OUT4$tagbim &gt; 1kG_SNPs.txt
plink --bfile $FILE_QC --extract 1kG_SNPs.txt --recode --make-bed --out $OUT1_QC</code></pre>
<em>The datasets now contain the exact same variants.</em></p>
<h2 id="change-build-on-1000-genomes-data-build-to-match-build-of-hapmap-data">Change build on 1000 Genomes data build to match build of HapMap data<a class="headerlink" href="#change-build-on-1000-genomes-data-build-to-match-build-of-hapmap-data" title="Permanent link">&para;</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Look at <a href="../liftover/">Liftover tutorial</a> to see how to move data set to another build. </p>
</div>
<pre class="highlight"><code class="language-bash">awk '{print$2,$4}' $OUT1_QC$tagmap &gt; buildmap.txt
# buildmap.txt contains one SNP-id and physical position per line.
plink --bfile $OUT4 --update-map buildmap.txt --make-bed --out $OUT5</code></pre>
<h2 id="merge-the-map-and-1000-genomes-data-sets">Merge the Map and 1000 Genomes data sets<a class="headerlink" href="#merge-the-map-and-1000-genomes-data-sets" title="Permanent link">&para;</a></h2>
<details class="note"><summary>Prior to merging 1000 Genomes data with the data we want to make sure that the files are mergeable, for this we conduct 3 steps:</summary><p>1) Make sure the reference genome is similar in your data and the 1000 Genomes Project datasets </p>
<p>2) Resolve strand issues. </p>
<p>3) Remove the SNPs which after the previous two steps still differ between datasets</p>
</details>
<p><strong>1) set reference genome</strong>
<pre class="highlight"><code class="language-bash">awk '{print$2,$5}' $OUT5$tagbim &gt; 1kg_ref-list.txt
plink --bfile $OUT1_QC --reference-allele 1kg_ref-list.txt --make-bed --out Map-adj
# The 1kG_MDS6 and the HapMap-adj have the same reference genome for all SNPs.</code></pre></p>
<p><strong>2) Resolve strand issues</strong>
<pre class="highlight"><code class="language-bash">awk '{print$2,$5,$6}' $OUT5$tagbim &gt; 1kGMDS_strand_tmp
awk '{print$2,$5,$6}' Map-adj.bim &gt; Map-adj_tmp
sort 1kGMDS_strand_tmp Map-adj_tmp |uniq -u &gt; all_differences.txt</code></pre></p>
<p><em>Flip SNPs for resolving strand issues</em>
<pre class="highlight"><code class="language-bash">awk '{print$1}' all_differences.txt | sort -u &gt; flip_list.txt
plink --bfile Map-adj --flip flip_list.txt --reference-allele 1kg_ref-list.txt --make-bed --out corrected_map</code></pre></p>
<p><em>Check for SNPs which are still problematic after they have been flipped.</em>
<pre class="highlight"><code class="language-bash">awk '{print$2,$5,$6}' corrected_map.bim &gt; corrected_map_tmp
sort 1kGMDS_strand_tmp corrected_map_tmp |uniq -u  &gt; uncorresponding_SNPs.txt</code></pre></p>
<p><strong>3) Remove problematic SNPs from your data and from the 1000 Genomes.</strong>
<pre class="highlight"><code class="language-bash">awk '{print$1}' uncorresponding_SNPs.txt | sort -u &gt; SNPs_for_exclusion.txt
plink --bfile corrected_map --exclude SNPs_for_exclusion.txt --make-bed --out $OUT2_QC
plink --bfile $OUT5 --exclude SNPs_for_exclusion.txt --make-bed --out $OUT6</code></pre></p>
<h2 id="merge-outdata-with-1000-genomes-data">Merge outdata with 1000 Genomes Data<a class="headerlink" href="#merge-outdata-with-1000-genomes-data" title="Permanent link">&para;</a></h2>
<pre class="highlight"><code class="language-bash">plink --bfile $OUT2_QC --bmerge $OUT6$tagbed $OUT6$tagbim $OUT6$tagfam --allow-no-sex --make-bed --out MDS_merge2</code></pre>
<h2 id="perform-mds-and-pca-on-map-ceu-data-anchored-by-1000-genomes-data-using-a-set-of-pruned-snps">Perform MDS and PCA on Map-CEU data anchored by 1000 Genomes data using a set of pruned SNPs<a class="headerlink" href="#perform-mds-and-pca-on-map-ceu-data-anchored-by-1000-genomes-data-using-a-set-of-pruned-snps" title="Permanent link">&para;</a></h2>
<p><strong>Download the file with population information of the 1000 genomes dataset.</strong>
<pre class="highlight"><code class="language-bash">wget ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20100804/20100804.ALL.panel</code></pre>
<em>The file 20130502.ALL.panel contains population codes of the individuals of 1000 genomes.</em></p>
<p><strong>Convert population codes into superpopulation codes (i.e., AFR,AMR,ASN, and EUR).</strong>
<pre class="highlight"><code class="language-bash">awk '{print$1,$1,$2}' 20100804.ALL.panel &gt; race_1kG.txt</code></pre></p>
<p><strong>Create a racefile of your own data.</strong>
<pre class="highlight"><code class="language-bash">awk '{print$1,$2,"OWN"}' $OUT2_QC$tagfam &gt; racefile_own.txt</code></pre>
<strong>make txt file of concatenated racefiles</strong>
<pre class="highlight"><code class="language-bash">sed 's/JPT/ASN/g' race_1kG.txt&gt;race_1kG2.txt
sed 's/ASW/AFR/g' race_1kG2.txt&gt;race_1kG3.txt
sed 's/CEU/EUR/g' race_1kG3.txt&gt;race_1kG4.txt
sed 's/CHB/ASN/g' race_1kG4.txt&gt;race_1kG5.txt
sed 's/CHD/ASN/g' race_1kG5.txt&gt;race_1kG6.txt
sed 's/YRI/AFR/g' race_1kG6.txt&gt;race_1kG7.txt
sed 's/LWK/AFR/g' race_1kG7.txt&gt;race_1kG8.txt
sed 's/TSI/EUR/g' race_1kG8.txt&gt;race_1kG9.txt
sed 's/MXL/AMR/g' race_1kG9.txt&gt;race_1kG10.txt
sed 's/GBR/EUR/g' race_1kG10.txt&gt;race_1kG11.txt
sed 's/FIN/EUR/g' race_1kG11.txt&gt;race_1kG12.txt
sed 's/CHS/ASN/g' race_1kG12.txt&gt;race_1kG13.txt
sed 's/PUR/AMR/g' race_1kG13.txt&gt;race_1kG14.txt
cat race_1kG14.txt racefile_own.txt | sed -e '1i\FID IID race' &gt; racefile.txt</code></pre></p>
<h1 id="perform-mds">Perform MDS<a class="headerlink" href="#perform-mds" title="Permanent link">&para;</a></h1>
<div class="tabbed-set" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">Perform MDS Clustering</label><div class="tabbed-content">
<pre class="highlight"><code class="language-bash">plink --bfile MDS_merge2 --extract $FILE_PRUNEIN --genome --out MDS_merge2
plink --bfile MDS_merge2 --read-genome MDS_merge2.genome --cluster --mds-plot 10 --out MDS_merge2</code></pre>
</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">Create MDS-plot</label><div class="tabbed-content">
<div class="tabbed-set" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">Performed in R</label><div class="tabbed-content"></div>
</div>
<pre class="highlight"><code>data&lt;- read.table(file="MDS_merge2.mds",header=TRUE)
race&lt;- read.table(file="racefile.txt",header=TRUE)
datafile&lt;- merge(data,race,by=c("IID","FID"))
head(datafile)

pdf("MDS.pdf",width=7,height=7)
for (i in 1:nrow(datafile))
{
if (datafile[i,14]=="EUR") {plot(datafile[i,4],datafile[i,5],type="p",xlim=c(-0.1,0.2),ylim=c(-0.15,0.1),xlab="MDS Component 1",ylab="MDS Component 2",pch=1,cex=0.5,col="green")}
par(new=T)
if (datafile[i,14]=="ASN") {plot(datafile[i,4],datafile[i,5],type="p",xlim=c(-0.1,0.2),ylim=c(-0.15,0.1),xlab="MDS Component 1",ylab="MDS Component 2",pch=1,cex=0.5,col="red")}
par(new=T)
if (datafile[i,14]=="AMR") {plot(datafile[i,4],datafile[i,5],type="p",xlim=c(-0.1,0.2),ylim=c(-0.15,0.1),xlab="MDS Component 1",ylab="MDS Component 2",pch=1,cex=0.5,col=470)}
par(new=T)
if (datafile[i,14]=="AFR") {plot(datafile[i,4],datafile[i,5],type="p",xlim=c(-0.1,0.2),ylim=c(-0.15,0.1),xlab="MDS Component 1",ylab="MDS Component 2",pch=1,cex=0.5,col="blue")}
par(new=T)
if (datafile[i,14]=="OWN") {plot(datafile[i,4],datafile[i,5],type="p",xlim=c(-0.1,0.2),ylim=c(-0.15,0.1),xlab="MDS Component 1",ylab="MDS Component 2",pch=3,cex=0.7,col="black")}
par(new=T)
}

abline(v=-0.035,lty=3)
abline(h=0.035,lty=3)
legend("topright", pch=c(1,1,1,1,3),c("EUR","ASN","AMR","AFR","OWN"),col=c("green","red",470,"blue","black"),bty="o",cex=1)
dev.off()</code></pre>
<p><img alt="MDS example" src="../img/MDS.png" /></p>
<blockquote>
<p>An example bar plot generated using script in <code>MDS_merged.R</code>    </p>
</blockquote>
</div>
</div>
<h1 id="perform-pca">Perform PCA<a class="headerlink" href="#perform-pca" title="Permanent link">&para;</a></h1>
<div class="tabbed-set" data-tabs="3:2"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">Perform PCA Clustering</label><div class="tabbed-content">
<pre class="highlight"><code class="language-bash">plink --bfile MDS_merge2 --indep-pairwise 50 5 0.5 
plink --bfile MDS_merge2 --extract plink.prune.in  --make-bed --pca 10 'header' --out PCA_MERGE</code></pre>
</div>
<input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><label for="__tabbed_3_2">Create PCA-plot</label><div class="tabbed-content">
<div class="tabbed-set" data-tabs="4:1"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">Performed in R</label><div class="tabbed-content"></div>
</div>
<p><pre class="highlight"><code>data&lt;- read.table(file="PCA_MERGE.eigenvec",header=TRUE)
race&lt;- read.table(file="racefile.txt",header=TRUE)
datafile&lt;- merge(data,race,by=c("IID","FID"))
head(datafile)

pdf("pca.pdf",width=7,height=7)
for (i in 1:nrow(datafile))
{
if (datafile[i,13]=="EUR") {plot(datafile[i,3],datafile[i,4],type="p",xlim=c(-0.05,0.1),ylim=c(-0.1,0.1),xlab="PC Component 1",ylab="PC Component 2",pch=1,cex=0.5,col="green")}
par(new=T)
if (datafile[i,13]=="ASN") {plot(datafile[i,3],datafile[i,4],type="p",xlim=c(-0.05,0.1),ylim=c(-0.1,0.1),xlab="PC Component 1",ylab="PC Component 2",pch=1,cex=0.5,col="red")}
par(new=T)
if (datafile[i,13]=="AMR") {plot(datafile[i,3],datafile[i,4],type="p",xlim=c(-0.05,0.1),ylim=c(-0.1,0.1),xlab="PC Component 1",ylab="PC Component 2",pch=1,cex=0.5,col="orange")}
par(new=T)
if (datafile[i,13]=="AFR") {plot(datafile[i,3],datafile[i,4],type="p",xlim=c(-0.05,0.1),ylim=c(-0.1,0.1),xlab="PC Component 1",ylab="PC Component 2",pch=1,cex=0.5,col="blue")}
par(new=T)
if (datafile[i,13]=="OWN") {plot(datafile[i,3],datafile[i,4],type="p",xlim=c(-0.05,0.1),ylim=c(-0.1,0.1),xlab="PC Component 1",ylab="PC Component 2",pch=3,cex=0.7,col="black")}
par(new=T)
}

abline(v=-0.005,lty=3)
abline(h=-0.005,lty=3)
legend("topright", pch=c(1,1,1,1,3),c("EUR","ASN","AMR","AFR","OWN"),col=c("green","red","orange","blue","black"),bty="o",cex=1)
dev.off()

## awk '{ if ($3 &lt;-0.005 &amp;&amp; $4 &gt;-0.005) print $1,$2 }' PCA_MERGE.eigenvec &gt; EUR_samp.txt</code></pre>
<img alt="PCA example" src="../img/pca.png" /></p>
</div>
</div>
<h2 id="run-admixture-algorithm">Run Admixture algorithm<a class="headerlink" href="#run-admixture-algorithm" title="Permanent link">&para;</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You must install the ADMIXTURE software <a href="http://dalexander.github.io/admixture/download.html">here</a>. If       you are running on a conda environment you can install the admixture software using the following command:
<pre class="highlight"><code class="language-bash">conda install -c bioconda admixture</code></pre></p>
</div>
<p><strong>Concatenate racefiles.</strong>
<pre class="highlight"><code class="language-bash">awk '{print$1,$2,"-"}' $OUT2_QC$tagfam &gt; racefile_own.txt
awk '{print$1,$1,$2}' 20100804.ALL.panel &gt; race_1kG.txt
cat racefile_own.txt race_1kG.txt| sed -e '1i\FID IID race' &gt; MDS_merge2.pop
sed -i -e "1d" MDS_merge2.pop
cut -d " " -f 3- MDS_merge2.pop &gt;temp.txt</code></pre>
<strong>make popfile for admixture script</strong>
<pre class="highlight"><code class="language-bash">mv temp.txt MDS_merge2.pop</code></pre></p>
<p><strong>run admixture script</strong>
<pre class="highlight"><code class="language-bash">qsub -cwd -pe smp 8 -l mem_free=32G -l scratch=100G -l h_rt=40:20:00 ad.sh
##admixture --supervised ./MDS_merge2.bed 12 &gt; log_merge_admixture.out</code></pre></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This step can take a very long time. Running overnight may be neccessary. </p>
</div>
<div class="tabbed-set" data-tabs="5:1"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><label for="__tabbed_5_1">Performed in R</label><div class="tabbed-content"></div>
</div>
<p><pre class="highlight"><code>R
tbl=read.table("MDS_merge2.12.Q")
popGroups = read.table("MDS_merge2.pop")
fam=read.table("MDS_merge2.fam")
mergedAdmWithPopGroups = cbind(tbl, popGroups)
ordered = mergedAdmWithPopGroups
#ceu v5
#tsi v12
#fin v2
#gbr v1
ordered$EUR=ordered$V5+ordered$V12+ordered$V2+ordered$V1
fam$race=ordered$EUR
ids1=fam$V1[which(fam$race&gt;0.8)]
ids=data.frame(ids1)
ids$ids2=fam$V2[which(fam$race&gt;0.8)]
##list of europeans to extract from files
write.table(ids,'europeans.txt',quote = FALSE,row.names = FALSE,col.names=FALSE)
ordered$V5=c()
ordered$V12=c()
ordered$V2=c()
ordered$V1=c()
ordered=ordered[order(-ordered$EUR),]
#barplot(t(as.matrix(subset(ordered, select=c("YRI","ASW","CHB","CHS","EUR","JPT","LWK","MXL","PUR")))), col=rainbow(12), border=NA,names.arg=popnew, las=2,cex.names=0.1)
pdf('admixture.pdf')
barplot(t(as.matrix(subset(ordered, select=c("V3" , "V4" , "V6" , "V7"  ,"V8"  ,"V9" ,"V10", "V11","EUR")))), col=rainbow(9), border=NA, las=2,ylab="Percent Pop",xaxt="n",xlab='&lt;---- Direction of more European')
dev.off()</code></pre>
<img alt="Admixture example" src="../img/Admix.png" /></p>
<blockquote>
<p>An example bar plot generated using script in <code>admixtureplot.R</code> </p>
</blockquote>
<p>The output of this script prints out a text file <strong>europeans.txt</strong> with a list of all the individuals that are found to be 80% or more European. </p>
<h2 id="exclude-ethnic-outliers">Exclude ethnic outliers.<a class="headerlink" href="#exclude-ethnic-outliers" title="Permanent link">&para;</a></h2>
<p><em>Select individuals in your own data below cut-off thresholds. The cut-off levels are not fixed thresholds but have to be determined based on the visualization of the first two dimensions. To exclude ethnic outliers, the thresholds need to be set around the cluster of population of interest.</em></p>
<p><pre class="highlight"><code class="language-bash">awk '{ if ($4 &lt;-0.04 &amp;&amp; $5 &gt;0.03) print $1,$2 }' MDS_merge2.mds &gt; EUR_MDS_merge2
plink --bfile $FILE_QC --keep EUR_MDS_merge2 --make-bed --out $FILEQCEURO</code></pre>
<em>to exclude ethnic outliers using output from ADMIXTURE use the following script</em>
<pre class="highlight"><code class="language-bash">plink --bfile $FILE_QC --keep europeans.txt --make-bed --out $FILEQCEURO</code></pre></p>
<h2 id="hardy-weinburg-equilibrium-filter-on-controls">Hardy weinburg equilibrium filter on controls<a class="headerlink" href="#hardy-weinburg-equilibrium-filter-on-controls" title="Permanent link">&para;</a></h2>
<pre class="highlight"><code class="language-bash">plink --bfile $FILEQCEURO --hwe $HWE_CONTROL --make-bed --out $OUTQCEURO</code></pre>
<h2 id="remove-all-relatedness-from-dataset">Remove all 'relatedness' from dataset<a class="headerlink" href="#remove-all-relatedness-from-dataset" title="Permanent link">&para;</a></h2>
<pre class="highlight"><code class="language-bash">plink --bfile $OUTQCEURO --filter-founders --make-bed --out $OUTQCEURO$sep$found</code></pre>
<div class="tabbed-set" data-tabs="6:1"><input checked="checked" id="__tabbed_6_1" name="__tabbed_6" type="radio" /><label for="__tabbed_6_1">Performed in R</label><div class="tabbed-content"></div>
</div>
<pre class="highlight"><code>data&lt;- read.table(file="kin.txt",header=TRUE)
data_related=data[,c(1,2,3,4)]
missing &lt;- read.table("plink.imiss", header =TRUE, as.is=T)
FID1=data_related[,c(1,2)]
FID2=data_related[,c(3,4)]

FID1$index=row.names(FID1)
FID2$index=row.names(FID2)

FID1$IID=FID1$IID1
FID2$IID=FID2$IID2

FID1[,1:2]=c()
FID2[,1:2]=c()

FID1=merge(FID1,missing,by="IID")
FID2=merge(FID2,missing,by="IID")

FID1[,4:6]=c()
FID2[,4:6]=c()

FID1$index=as.numeric(FID1$index)
FID2$index=as.numeric(FID2$index)

q=c(setdiff(FID2$index, FID1$index), setdiff(FID1$index, FID2$index))
if (length(q) != 0) {
FID1=FID1[!(FID1$index==q),]
FID2=FID2[!(FID2$index==q),]
}

FID1=FID1[order(FID1[,2]),]
FID2=FID2[order(FID2[,2]),]

##this binds the FIDs,IIDs,missingness for all pairs
bind=cbind(FID1,FID2)
bind$index=c()
##this displays just the values for missingness for each pair
bindval=cbind(FID1$F_MISS,FID2$F_MISS)
colnames(bindval)=c(1,2)
max=as.numeric(colnames(bindval)[apply(bindval,1,which.max)])

#finds the corresponding IID and FID (bind1 and bind2) for the individual with the higher missingness so it can be removed
bind1=bind[cbind(seq_along(max*3-1), max*3-1)]
bind2=bind[cbind(seq_along(max*3-2), max*3-2)]
final=cbind(bind1,bind2)
final=unique(final)
write.table(final, '0.2_low_call_rate.txt', append = FALSE, sep = " ", dec = ".",
            row.names = FALSE, col.names = FALSE,quote=FALSE)</code></pre>
<p><strong>remove individual with higher missingness</strong>
<pre class="highlight"><code class="language-bash">plink --bfile $OUTQCEURO$sep$found --remove 0.2_low_call_rate.txt --make-bed --out $OUTCOVER</code></pre></p>
<h2 id="create-covariates-based-on-mds">Create covariates based on MDS.<a class="headerlink" href="#create-covariates-based-on-mds" title="Permanent link">&para;</a></h2>
<p><em>Perform an MDS ONLY on qccase data without ethnic outliers. The values of the 10 MDS dimensions are subsequently used as covariates in the association analysis in the third tutorial</em>
<pre class="highlight"><code class="language-bash">plink --bfile $OUTCOVER --extract plink.prune.in --genome --out $OUTCOVER
plink --bfile $OUTCOVER --read-genome $OUTCOVER$taggenome --cluster --mds-plot 10 --out $POPSTRATOUT
awk '{print$1, $2, $4, $5, $6, $7,$8,$9,$10,$11,$12,$13}' $POPSTRATOUT$tag_m$tag_dot_mds &gt; covar_mds.txt</code></pre>
<strong>The values in covar_mds.txt will be used as covariates, to adjust for remaining population stratification, in the third tutorial where we will perform a genome-wide association analysis.</strong></p>
<pre class="highlight"><code class="language-bash">mv $OUTCOVER$tagbed ./popstratout.bed
mv $OUTCOVER$tagbim ./popstratout.bim
mv $OUTCOVER$tagfam ./popstratout.fam
cp popstratout* ../3_Association_GWAS
cp covar_mds.txt ../3_Association_GWAS
cd ../3_Association_GWAS</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../assoc/" class="btn btn-neutral float-right" title="3. Association Analyses">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../QC/" class="btn btn-neutral" title="1. Quality Control of GWAS Data"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../QC/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../assoc/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
