<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>3. Association Analyses - Basic Tutorial for Genome Wide Association Analysis</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "3. Association Analyses";
    var mkdocs_page_input_path = "assoc.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Genome Wide Association Analysis</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../QC/">1. Quality Control of GWAS Data</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../popstrat/">2. Population Stratification</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">3. Association Analyses</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#we-only-need-to-define-a-few-variables-for-this-section-of-the-analysis">We only need to define a few variables for this section of the analysis</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#perform-association-analyses">Perform Association Analyses:</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#account-for-multiple-testing">Account for Multiple testing</a>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing association results</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="" href="../assocdive.md">5. Deep dive into underpinnings of associations</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../Additional_considerations/">6. Additional Considerations</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../prs/">7.PRS</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Genome Wide Association Analysis</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>3. Association Analyses</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/assoc.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="performing-association-analyses">Performing Association Analyses<a class="headerlink" href="#performing-association-analyses" title="Permanent link">&para;</a></h1>
<p>In this tutorial, you will learn how to use the HapMap files generated in the <a href="../popstrat/">previous tutorial</a> named: popstratout (with .bed, .bim, and .fam. extensions) and covar_pca.txt to perform association analysis on a simulated phenotype. </p>
<h2 id="we-only-need-to-define-a-few-variables-for-this-section-of-the-analysis">We only need to define a few variables for this section of the analysis<a class="headerlink" href="#we-only-need-to-define-a-few-variables-for-this-section-of-the-analysis" title="Permanent link">&para;</a></h2>
<p>Before we start, we define two variable. <strong>FILE_GWAS</strong> refers to the dataset to perform the association analysis on and <strong>FILE_COV</strong> to the dataset with the relevant covariates (Principal components).</p>
<pre class="highlight"><code class="language-bash">FILE_GWAS=popstratout
FILE_COV=covar_PCs.txt</code></pre>
<h2 id="perform-association-analyses">Perform Association Analyses:<a class="headerlink" href="#perform-association-analyses" title="Permanent link">&para;</a></h2>
<p>We will use Plink to perform association analyses. These types of analyses run M individual regressions, with M the number of SNPs in the .bim file. For case/control data, the simples form of association analysis is performed using the <strong>--assoc</strong> flag. However, running multivariable regression analyses, one for each SNP, is an attractive way of estimating SNPs associations, correcting for a multitude of control variables. The <strong>--linear</strong> flag is used for linear regression (apropriate for quantitative traits and possibly binary traits), the <strong>--logistic</strong> flag for logistic regression (appropriate for binary traits). Below, we show how to perform each of these three types of association analyses on a binary, simulated phenotype (stored in the 6th column of the .fam file).  </p>
<p>All these analyses are performed M times, for each SNP. As a result, conventional p-values need to be adjusted for multiple hypothesis testing. The <strong>--adjust</strong> flag performs this correction, and will add different types of multiple hypothesis corrected p-values to the output.</p>
<div class="tabbed-set" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">--assoc</label><div class="tabbed-content">
<div class="admonition note">
<p class="admonition-title">Note</p>
</div>
<p>the --assoc option does not allow to correct covariates such as principal components (PC's) MDS components, or other variables such as sex     and age. We therefore do not recommend to use this type of anlysis, as results will be vulnerable to confounding by population    stratification.
<pre class="highlight"><code class="language-bash">plink --bfile $FILE_GWAS --assoc --adjust --out assoc_results</code></pre></p>
</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">logistic</label><div class="tabbed-content"></div>
</div>
<p>The <strong>--logistic</strong> method performs logistic analysis and allows you to include covariates in your association analysis. We include sex as a covariate (which is recommended for many phenotypes), adding sex to the <strong>--logistic</strong> flag as shown below. Sex will be inferred from the .fam file. We will be using 10 principal components as covariates in this logistic analysis. We use the PCA components calculated from the previous tutorial: covar_PCA.txt. In that tutorial, we showed in the scree plot that the first two PCs should in principal be sufficient to control for population stratification. Nonetheless, we add the first 10 as this is considered the minimum in the literature and controlling for too many PCs is unlikely to have a large effect on the overall results (Price et al., 2006). We use the option <strong>hide-covar</strong> to only show the additive results of the SNPs in the output file. </p>
<pre class="highlight"><code>```bash
plink --bfile $FILE_GWAS --covar $FILE_COV --logistic 'hide-covar' sex --adjust --out logistic_results
```
**Remove NA values, those might give problems generating plots in later steps.**
```bash 
awk '!/'NA'/' logistic_results.assoc.logistic &gt; logistic_results.assoc_2.logistic
```
</code></pre>
<div class="tabbed-set" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">--linear</label><div class="tabbed-content">
<p>In case of a quantitative outcome measure the option --logistic should be replaced by <strong>--linear</strong> as to perform linear regression analysis. The use of the --assoc option is also possible for quantitative outcome measures (as metioned previously, this option does not allow the use of covariates). </p>
</div>
</div>
<h2 id="account-for-multiple-testing">Account for Multiple testing<a class="headerlink" href="#account-for-multiple-testing" title="Permanent link">&para;</a></h2>
<p>There are various way to deal with multiple testing outside of the conventional genome-wide significance threshold of 5.0E-8, below we present a couple: </p>
<h1 id="adjust">adjust<a class="headerlink" href="#adjust" title="Permanent link">&para;</a></h1>
<p><pre class="highlight"><code class="language-bash">plink --bfile $FILE_GWAS -assoc --adjust --out adjusted_assoc_results
grep -oP '(?&lt;==).*' adjusted_assoc_results.log &gt;test.txt
sed 's/\.$//' test.txt &gt;lambda.txt</code></pre>
The output file gives a Bonferroni corrected p-value, along with FDR and others.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a computational intensive step. Further pros and cons of this method, which can be used for association and dealing with multiple testing, are described in our article corresponding to this tutorial (https://www.ncbi.nlm.nih.gov/pubmed/29484742).To reduce computational time, we only perform this test on a subset of the SNPs from chromosome 22. The EMP2 column provides the values for multiple testing corrected p-value.</p>
</div>
<pre class="highlight"><code class="language-bash"># Perform 1000000 permutations.
plink --bfile $FILE_GWAS --assoc --mperm 1000000 --out 1M_perm_result
# Order your data, from lowest to highest p-value.
sort -gk 4 1M_perm_result.assoc.mperm &gt; sorted_subset.txt
# Check ordered permutation results
head sorted_subset.txt</code></pre>
<div class="tabbed-set" data-tabs="3:2"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">Generating Manhattan plot in R</label><div class="tabbed-content">
<pre class="highlight"><code>list.of.packages &lt;- c("karyoploteR")
new.packages &lt;- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
library("karyoploteR")  
results_log &lt;- read.table("logistic_results.assoc_2.logistic", head=TRUE)
results_as &lt;- read.table("assoc_results.assoc", head=TRUE)
results_log=data.frame(results_log)
results_as=data.frame(results_as)
results_log$CHR &lt;- sub("^", "chr", results_log$CHR ) ## addchr to column
results_as$CHR &lt;- sub("^", "chr",results_as$CHR )
results_log=makeGRangesFromDataFrame(results_log,keep.extra.columns=TRUE,ignore.strand=TRUE,seqinfo=NULL,seqnames.field=c("seqnames", "seqname","chromosome", "chrom","chr", "chromosome_name","seqid"),start.field="BP",end.field=c("BP", "stop"),strand.field="strand",starts.in.df.are.0based=FALSE)
results_as=makeGRangesFromDataFrame(results_as,keep.extra.columns=TRUE,ignore.strand=TRUE,seqinfo=NULL,seqnames.field=c("seqnames", "seqname","chromosome", "chrom","chr", "chromosome_name","seqid"),start.field="BP",end.field=c("BP", "stop"),strand.field="strand",starts.in.df.are.0based=FALSE)
jpeg("manhattan.jpeg",width=800)
kp &lt;- plotKaryotype(plot.type=4, chromosomes=c("chr1", "chr2","chr3" ,"chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22"),labels.plotter = NULL)
kpAddChromosomeNames(kp,srt=45)
kpAddLabels(kp, labels = "results_as", srt=90, pos=3, r0=autotrack(1,2))
kp &lt;- kpPlotManhattan(kp, data=results_as,points.col = "2blues", r0=autotrack(1,2))
kpAddLabels(kp, labels = "results_log", srt=90, pos=3, r0=autotrack(2,2))
kp &lt;- kpPlotManhattan(kp, data=results_log,points.col = "2blues",  r0=autotrack(2,2))
dev.off()</code></pre>
</div>
<input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><label for="__tabbed_3_2">Generating QQ plot in R</label><div class="tabbed-content">
<pre class="highlight"><code>list.of.packages &lt;- c("ggplot2")
new.packages &lt;- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
library(ggplot2)
inflation=read.table("lambda.txt")
inflation=inflation$V1
rs_log &lt;- read.table("logistic_results.assoc_2.logistic", head=TRUE)
rs_as &lt;- read.table("assoc_results.assoc", head=TRUE)
###function for plotting with 95% confidence interval
gg_qqplot &lt;- function(ps, ci = 0.95) {
n  &lt;- length(ps)
df &lt;- data.frame(
  observed = -log10(sort(ps)),
  expected = -log10(ppoints(n)),
  clower   = -log10(qbeta(p = (1 - ci) / 2, shape1 = 1:n, shape2 = n:1)),
  cupper   = -log10(qbeta(p = (1 + ci) / 2, shape1 = 1:n, shape2 = n:1))
)
log10Pe &lt;- expression(paste("Expected -log"[10], plain(P)))
log10Po &lt;- expression(paste("Observed -log"[10], plain(P)))
ggplot(df) +
  geom_ribbon(
    mapping = aes(x = expected, ymin = clower, ymax = cupper),
    alpha = 0.1
  ) +
  geom_point(aes(expected, observed), shape = 1, size = 3) +
  geom_abline(intercept = 0, slope = 1, alpha = 0.5) +
  geom_line(aes(expected, cupper), linetype = 2, size = 0.5) +
  geom_line(aes(expected, clower), linetype = 2, size = 0.5) +
  xlab(log10Pe) +
  ylab(log10Po)
}
##Run--&gt;
jpeg("QQ-Plot_logistic_COVAR.jpeg")
gg_qqplot(rs_log$P) +
theme_bw(base_size = 24) +
annotate(
  geom = "text",
  x = -Inf,
  y = Inf,
  hjust = -0.15,
  vjust = 1 + 0.15 * 3,
  label = sprintf("Î» = %.2f", inflation),
  size = 8
) +
theme(
  axis.ticks = element_line(size = 0.5),
  panel.grid = element_blank()
  # panel.grid = element_line(size = 0.5, color = "grey80")
)
dev.off()
jpeg("QQ-Plot_assoc.jpeg")
gg_qqplot(rs_as$P) +
theme_bw(base_size = 24)+
theme(axis.ticks = element_line(size = 0.5),
  panel.grid = element_blank())
dev.off()</code></pre>
</div>
</div>
<p><img alt="Manhattan example" src="../img/manhattan.png" />
<img alt="QQ example" src="../img/qq.png" /></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../plink/" class="btn btn-neutral float-right" title="PLINK">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../popstrat/" class="btn btn-neutral" title="2. Population Stratification"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../popstrat/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../plink/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
