<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>1. Quality Control of GWAS Data - Basic Tutorial for Genome Wide Association Analysis</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "1. Quality Control of GWAS Data";
    var mkdocs_page_input_path = "QC.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Genome Wide Association Analysis</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">1. Quality Control of GWAS Data</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#obtaining-and-visualizing-missingness-in-the-hapmap-data-files">Obtaining and visualizing missingness in the HapMap data files</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#reading-the-genotype-data-file-and-produce-missingness-data">Reading the genotype data file and produce missingness data:</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#generate-histogram-to-view-missingness">Generate Histogram to View Missingness.</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#filtering-steps">Filtering steps</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#filter-on-missingness">Filter on missingness</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#check-for-sex-discrepancies">Check for Sex Discrepancies</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#limit-data-to-autosomal-snps-only">Limit data to autosomal SNPs only</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#perform-maf-check-and-filter-by-maf-threshold">Perform MAF check and filter by MAF Threshold</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#view-heterozygosity-distribution-and-remove-outliers">View heterozygosity distribution and remove outliers</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#transfer-files-to-use-for-population-stratification">Transfer files to use for population stratification</a>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../popstrat/">2. Population Stratification</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../cal_prs/">3. Association Analyses</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing association results</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="" href="../assocdive.md">5. Deep dive into underpinnings of associations</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Genome Wide Association Analysis</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>1. Quality Control of GWAS Data</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/QC.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="qc-of-hapmap3-data">QC of HapMap3 data<a class="headerlink" href="#qc-of-hapmap3-data" title="Permanent link">&para;</a></h1>
<h2 id="obtaining-and-visualizing-missingness-in-the-hapmap-data-files">Obtaining and visualizing missingness in the HapMap data files<a class="headerlink" href="#obtaining-and-visualizing-missingness-in-the-hapmap-data-files" title="Permanent link">&para;</a></h2>
<p>The first step in GWAS analyses is to generate or obtain the genotype or sequencing data. Ideally these will be in Plink format (.bim,.bed,.fam)</p>
<p>You can download the HapMap example data <a href="https://drive.google.com/drive/folders/1TNRwAxWx_itjjt3dbtsR0CrUgkKk5yxw?usp=sharing">here</a> 
You <strong>MUST</strong> download and unzip the R folder from <a href="https://drive.google.com/drive/folders/1TNRwAxWx_itjjt3dbtsR0CrUgkKk5yxw?usp=sharing">here</a> as well to run downstream scripts. </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Due to limitation to bandwidth, we are currently using google drive to host the files, which doesn't allow the use of wget or curl to download the file. Please download the files manually.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you download the summary statistics on a MAC machine, the gz file may decompress automatically, which might result in a corrupted Height.gwas.txt file (including only the header). If this happens then you'll need to disable automatic decompression when downloading the file (eg. follow instructions <a href="https://octet.oberlin.edu/does-your-mac-unzip-zip-files-automatically/">here</a>)</p>
</div>
<h3 id="reading-the-genotype-data-file-and-produce-missingness-data">Reading the genotype data file and produce missingness data:<a class="headerlink" href="#reading-the-genotype-data-file-and-produce-missingness-data" title="Permanent link">&para;</a></h3>
<p>We will set a fileset variable to make it easier to adjust all downstream code and then produce SNP and individual missingness data:</p>
<pre class="highlight"><code class="language-bash">FILESET=HapMap_3_r3_1
plink -bfile $FILESET --missing </code></pre>
<p>This will print a <strong>plink.imiss</strong> and <strong>plink.lmiss</strong> output file with the individual and SNP missingness data, respectively.  </p>
<p>The <strong>plink.imiss</strong> file contains the following columns:</p>
<table>
<thead>
<tr>
<th align="center">FID</th>
<th align="center">IID</th>
<th align="center">MISS_PHENO</th>
<th align="center">N_MISS</th>
<th align="center">N_GENO</th>
<th align="center">F_MISS</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">1328</td>
<td align="center">NA06989</td>
<td align="center">N</td>
<td align="center">4203</td>
<td align="center">1457897</td>
<td align="center">0.002883</td>
</tr>
<tr>
<td align="center">1377</td>
<td align="center">NA11891</td>
<td align="center">N</td>
<td align="center">20787</td>
<td align="center">1457897</td>
<td align="center">0.01426</td>
</tr>
<tr>
<td align="center">1349</td>
<td align="center">NA11843</td>
<td align="center">N</td>
<td align="center">1564</td>
<td align="center">1457897</td>
<td align="center">0.001073</td>
</tr>
</tbody>
</table>
<p>The column headers correspond to the following: </p>
<ul>
<li><strong>FID</strong>: Family ID</li>
<li><strong>IID</strong>: Individual ID</li>
<li><strong>MISS_PHENO</strong>:  Missing phenotype? (Y/N)</li>
<li><strong>N_MISS</strong>: Number of missing SNPs</li>
<li><strong>N_GENO</strong>: Number of non-obligatory missing genotypes</li>
<li><strong>F_MISS</strong>: Proportion of missing SNPs</li>
</ul>
<h3 id="generate-histogram-to-view-missingness">Generate Histogram to View Missingness.<a class="headerlink" href="#generate-histogram-to-view-missingness" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="1:1"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">Performed in R</label><div class="tabbed-content"></div>
</div>
<pre class="highlight"><code>indmiss&lt;-read.table(file="plink.imiss", header=TRUE)
snpmiss&lt;-read.table(file="plink.lmiss", header=TRUE)
# read data into R 

png("histimiss.png") #indicates pdf format and gives title to file
hist(indmiss[,6],main="Histogram individual missingness") #selects column 6, names header of file
dev.off()
png("histlmiss.png")
hist(snpmiss[,5],main="Histogram SNP missingness")  
dev.off() # shuts down the current device</code></pre>
<p><img alt="Example Bar Plot" src="../img/histimiss.png" />
<img alt="Example Bar Plot" src="../img/histlmiss.png" /></p>
<blockquote>
<p>An example bar plot generated using script in <code>hist_miss.R</code>    </p>
</blockquote>
<h2 id="filtering-steps">Filtering steps<a class="headerlink" href="#filtering-steps" title="Permanent link">&para;</a></h2>
<h4 id="filter-on-missingness">Filter on missingness<a class="headerlink" href="#filter-on-missingness" title="Permanent link">&para;</a></h4>
<p>We want to begin performing quality control filtering on the data.</p>
<p>We begin with organizing our file naming convention and define our starting filter values. </p>
<p><pre class="highlight"><code class="language-bash">sep="_"
TAG_GENO="geno"
GENO=0.02 
TAG_MIND="mind"
MIND=0.02 
OUT="$FILESET$sep$TAG_GENO"
OUT1="$FILESET$sep$TAG_GENO$sep$TAG_MIND"</code></pre>
In is <strong>important</strong> to filter on SNP missingness (--geno) prior to filtering on individual missingness (--mind). </p>
<pre class="highlight"><code class="language-bash">plink --bfile $FILESET --geno $GENO --make-bed --out $OUT
plink --bfile $OUT --mind $MIND --make-bed --out $OUT1</code></pre>
<h4 id="check-for-sex-discrepancies">Check for Sex Discrepancies<a class="headerlink" href="#check-for-sex-discrepancies" title="Permanent link">&para;</a></h4>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>if you have not already done so, YOU MUST use the split-x command to remove the psuedo autosomal region</p>
</div>
<div class="tabbed-set" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">Performed in R</label><div class="tabbed-content"></div>
</div>
<pre class="highlight"><code>gender &lt;- read.table("plink.sexcheck", header=T,as.is=T)

pdf("Gender_check.pdf")
hist(gender[,6],main="Gender", xlab="F")
dev.off()

pdf("Men_check.pdf")
male=subset(gender, gender$PEDSEX==1)
hist(male[,6],main="Men",xlab="F")
dev.off()

pdf("Women_check.pdf")
female=subset(gender, gender$PEDSEX==2)
hist(female[,6],main="Women",xlab="F")
dev.off()</code></pre>
<p><img alt="Example Bar Plot" src="../img/gender.png" /></p>
<p>Remove SNPS that do not pass the sex check. </p>
<pre class="highlight"><code class="language-bash">rem="rem"
OUT2=$OUT1$sep$rem
# This command generates a list of individuals with the status ?PROBLEM?.
grep "PROBLEM" plink.sexcheck| awk '{print$1,$2}'&gt; sex_discrepancy.txt
plink --bfile $OUT1 --remove sex_discrepancy.txt --make-bed --out $OUT2</code></pre>
<h4 id="limit-data-to-autosomal-snps-only">Limit data to autosomal SNPs only<a class="headerlink" href="#limit-data-to-autosomal-snps-only" title="Permanent link">&para;</a></h4>
<p>We wll generate a txt file with all the SNPS on the autosomes so that we can remove all SNPS that are on the sex chromosomes or in mitochondrial DND. 
<pre class="highlight"><code class="language-bash">BIM=".bim"
AUTOSOME="autosome"
OUT3=$OUT2$sep$AUTOSOME

awk '{ if ($1 &gt;= 1 &amp;&amp; $1 &lt;= 23) print $2 }' $OUT2$BIM &gt; snp_1_22.txt
plink --bfile $OUT2 --extract snp_1_22.txt --make-bed --out $OUT3</code></pre></p>
<h4 id="perform-maf-check-and-filter-by-maf-threshold">Perform MAF check and filter by MAF Threshold<a class="headerlink" href="#perform-maf-check-and-filter-by-maf-threshold" title="Permanent link">&para;</a></h4>
<pre class="highlight"><code class="language-bash">plink --bfile $OUT2 --freq --out MAF_check</code></pre>
<div class="tabbed-set" data-tabs="3:1"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">Performed in R</label><div class="tabbed-content"></div>
</div>
<p><pre class="highlight"><code>maf_freq &lt;- read.table("MAF_check.frq", header =TRUE, as.is=T)
pdf("MAF_distribution.pdf")
hist(maf_freq[,5],main = "MAF distribution", xlab = "MAF")
dev.off()</code></pre>
<img alt="Example" src="../img/MAF.png" /></p>
<p>Now we want to filter out any SNPS with a MAF&lt;0.05 for a small data set or 0.1 for a larger data set. </p>
<pre class="highlight"><code class="language-bash">TAG_MAF="maf"
OUT3=$OUT2$sep$TAG_MAF
MAF=0.05
plink --bfile $OUT2 --maf $MAF --make-bed --out $OUT3</code></pre>
<h4 id="view-heterozygosity-distribution-and-remove-outliers">View heterozygosity distribution and remove outliers<a class="headerlink" href="#view-heterozygosity-distribution-and-remove-outliers" title="Permanent link">&para;</a></h4>
<p><strong>Generate a pruned subset of SNPs that are in approximate linkage equilibrium.</strong>
<pre class="highlight"><code class="language-bash">plink --bfile $OUT3 --indep-pairwise 50 5 0.5</code></pre>
<strong>Compute method-of-moments F coefficient estimates</strong></p>
<p>The --het flag in Plink computes observed and expected autosomal homozygous genotype counts for each sample, and reports method-of-moments F coefficient estimates (i.e. (<observed hom. count> - <expected count>) / (<total observations> - <expected count>)) to R_check.het. </p>
<pre class="highlight"><code class="language-bash">plink --bfile $OUT3 --extract plink.prune.in --het --out R_check </code></pre>
<div class="tabbed-set" data-tabs="4:1"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">Plot of the heterozygosity rate distribution generated in R</label><div class="tabbed-content"></div>
</div>
<p><pre class="highlight"><code>het &lt;- read.table("R_check.het", head=TRUE)
pdf("heterozygosity.pdf")
het$HET_RATE = (het$"N.NM." - het$"O.HOM.")/het$"N.NM."
hist(het$HET_RATE, xlab="Heterozygosity Rate", ylab="Frequency", main= "Heterozygosity Rate")
dev.off()</code></pre>
<img alt="Example" src="../img/het.png" /></p>
<div class="tabbed-set" data-tabs="5:1"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><label for="__tabbed_5_1">list of individuals who deviate more than 3 standard deviations from the heterozygosity rate mean generated in R</label><div class="tabbed-content"></div>
</div>
<pre class="highlight"><code>het &lt;- read.table("R_check.het", head=TRUE)
het$HET_RATE = (het$"N.NM." - het$"O.HOM.")/het$"N.NM."
het_fail = subset(het, (het$HET_RATE &lt; mean(het$HET_RATE)-3*sd(het$HET_RATE)) | (het$HET_RATE &gt; mean(het$HET_RATE)+3*sd(het$HET_RATE)));
het_fail$HET_DST = (het_fail$HET_RATE-mean(het$HET_RATE))/sd(het$HET_RATE);
write.table(het_fail, "fail-het-qc.txt", row.names=FALSE)</code></pre>
<p>The command above generates the <strong>fail-het-qc.txt</strong> file and the command below saves the first two rows to a new file named <strong>het_fail_ind.txt</strong></p>
<p><pre class="highlight"><code class="language-bash">sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}' &gt; het_fail_ind.txt</code></pre>
<strong>remove individuals that are heterozygous outliers</strong>
<pre class="highlight"><code class="language-bash">HET="het_fix"
OUT4=$OUT3$sep$HET
plink --bfile $OUT3 --remove het_fail_ind.txt --make-bed --out $OUT4</code></pre></p>
<h2 id="transfer-files-to-use-for-population-stratification">Transfer files to use for population stratification<a class="headerlink" href="#transfer-files-to-use-for-population-stratification" title="Permanent link">&para;</a></h2>
<p><strong>rename files and copy to population stratification folder</strong> 
<pre class="highlight"><code class="language-bash">mkdir 2_Population_stratification
bed=".bed"
fam=".fam"
bim=".bim"
mv $OUT4$bed ./qcout.bed
mv $OUT4$bim ./qcout.bim
mv $OUT4$fam ./qcout.fam
cp qcout* 2_Population_stratification
cp plink.prune.in 2_Population_stratification</code></pre>
<strong>change directory into population stratification folder</strong>
<pre class="highlight"><code class="language-bash">cd 2_Population_stratification</code></pre></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../popstrat/" class="btn btn-neutral float-right" title="2. Population Stratification">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../popstrat/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
